wolfcrypt/src/port/Renesas/renesas_tsip_util.c

```diff
@@ -0,0 +1,654 @@
+                                                                    innerSz);
+    
+    if (ret == TSIP_SUCCESS)
+        ret = R_TSIP_Sha256HmacGenerateUpdate(&_handle, (uint8_t*)in, sz);
+    
+    if (ret == TSIP_SUCCESS)
+        ret = R_TSIP_Sha256HmacGenerateFinal(&_handle, digest);
+    
+    /* unlock hw */
+    tsip_hw_unlock();
+    
+    return ret;
+}
+
+/* generate Verify Data based on master secret */
+int tsip_generateVerifyData(const byte *ms, /* master secret */
+                            const byte *side, const byte *handshake_hash,
+                            byte *hashes /* out */)
+{
+    int ret ;
+    uint32_t l_side = R_TSIP_TLS_GENERATE_CLIENT_VERIFY;
+    
+    if (XSTRNCMP((const char*)side, (const char*)tls_server, FINISHED_LABEL_SZ)
+                                                                           == 0)
+    {
+        l_side = R_TSIP_TLS_GENERATE_SERVER_VERIFY;
+    }
+    
+    if ((ret = tsip_hw_lock()) == 0) {
+        ret = R_TSIP_TlsGenerateVerifyData(l_side, (uint32_t*)ms,
+                       (uint8_t*)handshake_hash, hashes/* out */);
+        if (ret != TSIP_SUCCESS) {
+            WOLFSSL_MSG("R_TSIP_TlsGenerateSessionKey failed\n");
+        }
+    }
+    /* unlock hw */
+    tsip_hw_unlock();
+    
+    return ret;
+}
+
+/* generate keys for TLS communication */
+int tsip_generateSeesionKey(struct WOLFSSL *ssl)
+{
+    int ret;
+    Ciphers *enc;
+    Ciphers *dec;
+    tsip_hmac_sha_key_index_t key_client_mac;
+    tsip_hmac_sha_key_index_t key_server_mac;
+    tsip_aes_key_index_t key_client_aes;
+    tsip_aes_key_index_t key_server_aes;
+    
+    if ((ret = tsip_hw_lock()) == 0) {
+        ret = R_TSIP_TlsGenerateSessionKey( _tls2tsipdef(ssl->options.cipherSuite),
+                (uint32_t*)ssl->arrays->tsip_masterSecret, 
+                (uint8_t*)ssl->arrays->clientRandom,
+                (uint8_t*)ssl->arrays->serverRandom, &key_client_mac,
+                &key_server_mac, &key_client_aes, &key_server_aes,
+                NULL, NULL);
+        if (ret != TSIP_SUCCESS) {
+            WOLFSSL_MSG("R_TSIP_TlsGenerateSessionKey failed\n");
+        } else {
+            /* successed creating session keys */
+            /* alloc aes instance for both enc and dec */
+            enc = &ssl->encrypt;
+            dec = &ssl->decrypt;
+            
+            if (enc) {
+                if (enc->aes == NULL)
+                    enc->aes = (Aes*)XMALLOC(sizeof(Aes), ssl->heap, 
+                                                    DYNAMIC_TYPE_CIPHER);
+                if (enc->aes == NULL)
+                    return MEMORY_E;
+                XMEMSET(enc->aes, 0, sizeof(Aes));
+            }
+            if (dec) {
+                if (dec->aes == NULL)
+                    dec->aes = (Aes*)XMALLOC(sizeof(Aes), ssl->heap, 
+                                                    DYNAMIC_TYPE_CIPHER);
+                if (dec->aes == NULL)
+                    return MEMORY_E;
```

Comment:
Do we need to free "enc->aes" here if this allocation for "dec->aes" fails?

Reply:
Yes. I added the code for freeing the memory in the case of allocation failure.